src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -282,5 +283,73 @@ private static IEnumerable<JsonConverter> DefaultSimpleConverters
                 yield return new UriConverter();
             }
         }
+
+        // Get converter from the cached key converters or add a new Enum converter.
+        internal JsonConverter GetOrAddKeyConverter(Type keyType, Type dictionaryType)
+        {
+            if (s_keyConverters.TryGetValue(keyType, out JsonConverter? converter))
+            {
+                return converter;
+            }
+            // short-term solution, instead of using a factory pattern like JsonStringEnumConverter,
+            // we just create the enum converter here add we add it to s_keyConverters.
+            else if (keyType.IsEnum)
+            {
+                converter = CreateEnumKeyConverter(keyType);
+                // Ignore failure case here in multi-threaded cases since the cached item will be equivalent.
+                s_keyConverters.TryAdd(keyType, converter);
+
+                return converter;
+            }
+            else
+            {
+                ThrowHelper.ThrowNotSupportedException_SerializationNotSupported(dictionaryType);
+                return null;
+            }
+        }
+
+        private JsonConverter CreateEnumKeyConverter(Type type)
+        {
+            JsonConverter converter = (JsonConverter)Activator.CreateInstance(
+                typeof(EnumKeyConverter<>).MakeGenericType(type),
+                BindingFlags.Instance | BindingFlags.Public,
+                binder: null,
+                Array.Empty<object>(),
+                culture: null)!;
+
+            return converter;
+        }
+
+        // The global list of built-in key converters.
+        private static readonly Dictionary<Type, JsonConverter> s_keyConverters = GetSupportedKeyConverters();
```

Comment:
We use a `ConcurrentDictionary` for caching the regular converters, but I tried with a plain dictionary and got the same result, could you explain why do we use the `ConcurrentDictionary` [there](https://github.com/dotnet/runtime/blob/5426b70e5309f5005ce47a33b2e0daaeb57a5ebd/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs#L26)?
@steveharter @layomia

Reply:
Its for thread safety: https://stackoverflow.com/questions/42013226/when-should-i-use-concurrentdictionary-and-dictionary

Consider adding tests to test multi-threaded access to this dictionary to verify things don't breaks (maybe register different converters for the same type and see if out-of-order additions from different threads causes undefined behavior).

Reply:
I cannot register different converters for this specific `KeyConverter` scenario... I guess that is why the plain Dictionary does not show an issue, however, the docs say that `ConcurrentDictionary` is faster on multi-threads so that would be a fair point to switch.

Reply:
Global state (i.e. object held in static variables) can be accessed by multiple threads, so we must use `ConcurrentDictionary`.